File: python/avin/domain/ticker.py
# ────────────────────────────────────────────────────────────────────────────
#  AVIN
#  Understand the market before trading it.
#
#  https://avin.info
# ────────────────────────────────────────────────────────────────────────────

from __future__ import annotations

from avin._native import PyTicker


class Ticker:
    """
    Trading instrument ticker.

    A ticker must be non-empty and must not contain whitespace.

    Parameters
    ----------
    value : str
        Instrument ticker.

    Raises
    ------
    ValueError
        If the ticker is empty or contains whitespace.

    Examples
    --------
    >>> ticker = Ticker("SBER")
    >>> str(ticker)
    'SBER'

    >>> Ticker("SiU6") == Ticker("SiU6")
    True
    """

    __slots__ = ("_inner",)

    def __init__(self, value: str) -> None:
        self._inner = PyTicker(value)

    def __str__(self) -> str:
        return self._inner.display()

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Ticker):
            return NotImplemented

        return self._inner.eq(other._inner)

    @classmethod
    def _from_native(cls, inner: PyTicker) -> Ticker:
        obj = cls.__new__(cls)
        obj._inner = inner

        return obj

File: crates/avin_domain/src/instrument/ticker.rs
// ───────────────────────────────────────────────────────────────────────────
// AVIN
// Understand the market before trading it.
//
// https://avin.info
// ───────────────────────────────────────────────────────────────────────────

use std::fmt::Display;

use crate::DomainError;

/// Trading instrument ticker.
///
/// For example: `"SBER"`, `"SiU6"`, `"BTCUSDT"`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Ticker(String);

impl Ticker {
    /// Creates a trading instrument ticker.
    ///
    /// # Errors
    ///
    /// Returns an error if the ticker is empty or contains whitespace.
    pub fn new(value: impl Into<String>) -> Result<Self, DomainError> {
        let s: String = value.into();

        validate_ticker(&s)?;

        Ok(Self(s))
    }
}

impl Display for Ticker {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

fn validate_ticker(s: &str) -> Result<(), DomainError> {
    if s.is_empty() {
        return Err(DomainError::Ticker(
            "instrument ticker can't be empty".to_string(),
        ));
    }

    if s.chars().any(|c| c.is_whitespace()) {
        return Err(DomainError::Ticker(
            "instrument ticker can't contain whitespace".to_string(),
        ));
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_ticker() {
        assert!(Ticker::new("SBER").is_ok());
        assert!(Ticker::new("SiU6").is_ok());
        assert!(Ticker::new("BTCUSDT").is_ok());
    }

    #[test]
    fn empty_ticker() {
        assert!(matches!(
            Ticker::new("").unwrap_err(),
            DomainError::Ticker(_)
        ));
    }

    #[test]
    fn has_whitespace() {
        assert!(matches!(
            Ticker::new(" ").unwrap_err(),
            DomainError::Ticker(_)
        ));
        assert!(matches!(
            Ticker::new("SBER ").unwrap_err(),
            DomainError::Ticker(_)
        ));
        assert!(matches!(
            Ticker::new(" SBER").unwrap_err(),
            DomainError::Ticker(_)
        ));
        assert!(matches!(
            Ticker::new("SB ER").unwrap_err(),
            DomainError::Ticker(_)
        ));
    }

    #[test]
    fn display() {
        let ticker = Ticker::new("SBER").unwrap();
        assert_eq!(ticker.to_string(), "SBER");
    }
}
